File: leetcode/1_two_sum.cpp
/*
 * @lc app=leetcode id=1 lang=cpp
 *
 * [1] Two Sum
 */

#include <vector>
#include <map>

// @lc code=start
class Solution {
 public:
  std::vector<int> twoSum(std::vector<int>& nums, int target) {
    std::map<int, int> m;

    for (auto i = 0; i < nums.size(); i++) {
      int num = nums[i];
      int diff = target - num;
      auto it = m.find(diff);
      if (it != m.end()) {
        return {it->second, i};
      }
      m[num] = i;
    }

    return {};
  }
};
// @lc code=end

File: src/1_two_sum.go
/*
 * @lc app=leetcode id=1 lang=golang
 *
 * [1] Two Sum
 */

package leetcode

// @lc code=start
func twoSum(nums []int, target int) []int {
	m := make(map[int]int)

	for i, num := range nums {
		diff := target - num
		if index, ok := m[diff]; ok {
			return []int{index, i}
		}
		m[num] = i
	}

	return nil
}

// @lc code=end
